### PR TITLE
support unregister cluster in karmada

### PR DESCRIPTION
**Pull request by wulemao**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

support unregister cluster in karmada

**Which issue(s) this PR fixes**:

Part of #5380

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
karmadactl: support unregister a pull mode cluster
```

### CODE REVIEW

**Comment by wulemao** [comment]
latest test report:

图片

**Comment by wulemao** [comment]
According to what we communicated offline, pr is updated

@chaosi-zju  ready to review

File: pkg/karmadactl/util/cluster.go
Comment (wulemao) [reply pkg/karmadactl/util/cluster.go:33]: yes

File: pkg/karmadactl/register/register.go
Comment (wulemao) [reply pkg/karmadactl/register/register.go:94]: actually no any sensitive information

`gosec` nolint is required just because the variable name `KarmadaConfigSecretName` contains the word `Secret`.

However, I don't want to rename to other name, I think the current variable name is the most accurate.

File: artifacts/deploy/bootstrap-token-configuration.yaml
Comment (wulemao) [reply artifacts/deploy/bootstrap-token-configuration.yaml:111]: Actually, this `ClusterrRole` is bound to group `system:nodes`, and only useful to `karmada-agent` which is installed by `karmadactl register`, it needs to request a certificate with the group `system:nodes` from `karmada-kube-controller-manager` to access Karmada.

`delete` permission is required because `karmadactl unregister` should use that `karmada-agent-config` to access Karmada to delete `cluster` object.

> PS: if `karmada-agent` is installed by `local up` or `helm`, it use certificate with group `system:masters` to access Karmada, not using this clusterrole., because the certificate is from `karmada.config` we passed by secret.

File: pkg/karmadactl/unregister/unregister.go
Comment (zhzhuang-zju) [inline pkg/karmadactl/unregister/unregister.go:148]:
```suggestion
	if errMsgs := validation.ValidateClusterName(j.ClusterName); len(errMsgs) != 0 {
		return fmt.Errorf("invalid cluster name(%s): %s", j.ClusterName, strings.Join(errMsgs, ";"))
	}
```

File: pkg/karmadactl/unregister/unregister.go
Comment (zhzhuang-zju) [inline pkg/karmadactl/unregister/unregister.go:226]: Can this be extracted into a method? Like `func(j *CommandUnregisterOption) getClusterResources() []obj`

File: pkg/karmadactl/unregister/unregister.go
Comment (zhzhuang-zju) [inline pkg/karmadactl/unregister/unregister.go:228]:
```suggestion
		// 2.2 delete karmada-agent deployment, its own rbac resource and secret/karmada-agent-config
```

File: pkg/karmadactl/unregister/unregister.go
Comment (zhzhuang-zju) [inline pkg/karmadactl/unregister/unregister.go:262]: https://github.com/karmada-io/karmada/blob/c6c20a9e22c3bb96ca487b388318712b7dc4d0c3/pkg/karmadactl/register/register.go#L343-L351
Will file `register.KarmadaAgentBootstrapKubeConfigFileName` be deleted after `karmadactl register` completes?

Comment (wulemao) [reply pkg/karmadactl/unregister/unregister.go:262]: It looks like it will

**Review by RainbowMango — COMMENTED** [review]
/assign

File: artifacts/deploy/bootstrap-token-configuration.yaml
Comment (zhzhuang-zju) [reply artifacts/deploy/bootstrap-token-configuration.yaml:111]: @B1F030 The RBAC permissions for `karmada-agent` need to be updated after this PR is merged.

File: pkg/karmadactl/unregister/unregister.go
Comment (wulemao) [reply pkg/karmadactl/unregister/unregister.go:226]: thanks, but what are the benefits? Isn't this clearer now?

Comment (zhzhuang-zju) [reply pkg/karmadactl/unregister/unregister.go:226]: Extracting this part of the code would help make the `RunUnregisterCluster` process clearer, preventing readers from being blocked by a large list of resources when reading the code. It would also make it easier to write unit tests. Of course, this is just my suggestion.

**Comment by wulemao** [comment]
updated, latest test result:

图片

is there any more comments?

**Comment by wulemao** [comment]
图片

updated, thank you for checking again.

File: pkg/karmadactl/unregister/unregister.go
Comment (zhzhuang-zju) [inline pkg/karmadactl/unregister/unregister.go:61]: indentation here seems have some issues

**Comment by zhzhuang-zju** [comment]
thanks~ others LGTM

File: pkg/karmadactl/util/cluster.go
Comment (RainbowMango) [inline pkg/karmadactl/util/cluster.go:32]:
```suggestion
// DeleteClusterObject deletes the cluster object from the Karmada control plane.
```

File: pkg/karmadactl/unregister/unregister.go
Comment (RainbowMango) [inline pkg/karmadactl/unregister/unregister.go:160]:
```suggestion
// Validate ensures that command unregister options are valid.
```

File: pkg/karmadactl/unregister/unregister.go
Comment (RainbowMango) [inline pkg/karmadactl/unregister/unregister.go:303]:
```suggestion
	// 1. delete the cluster object from the Karmada control plane
```

File: pkg/karmadactl/unregister/unregister.go
Comment (RainbowMango) [reply pkg/karmadactl/unregister/unregister.go:61]: All lines start with `Tab` instead of `space`, right?

**Review by RainbowMango — COMMENTED** [review]
Just some nits, otherwise LGTM.

File: pkg/karmadactl/unregister/unregister.go
Comment (wulemao) [reply pkg/karmadactl/unregister/unregister.go:61]: I don't get it. Can you explain it again?

Comment (zhzhuang-zju) [reply pkg/karmadactl/unregister/unregister.go:61]: please refer to https://github.com/karmada-io/karmada/blob/2a4567cbcc9a305baf977ba66133b9c55108347b/pkg/karmadactl/register/register.go#L68-L74

Comment (wulemao) [reply pkg/karmadactl/unregister/unregister.go:61]: 图片

Comment (wulemao) [reply pkg/karmadactl/unregister/unregister.go]: done

**Comment by wulemao** [comment]
comments fixed, any more comments? @RainbowMango

File: pkg/karmadactl/unregister/unregister.go
Comment (wulemao) [reply pkg/karmadactl/unregister/unregister.go]: all done

**Comment by zhzhuang-zju** [comment]
https://github.com/karmada-io/karmada/blob/2a4567cbcc9a305baf977ba66133b9c55108347b/pkg/controllers/certificate/cert_rotation_controller.go#L52-L60

@wulemao @chaosi-zju 
`CertRotationController` in  `karmad-agent` will look up and update the secret `karmada-kubeconfig`, which has been renamed to `karmada-agent-config`. We can use `func getKarmadaAgentConfig` to optimize this hard-coded section.

**Comment by wulemao** [comment]
updated, latest test result:

图片

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
